File: python/yacht/yacht.py
# Score categories.
# Change the values as you see fit.
YACHT = 'YACHT'
ONES = 'ONES'
TWOS = 'TWOS'
THREES = 'THREES'
FOURS = 'FOURS'
FIVES = 'FIVES'
SIXES = 'SIXES'
FULL_HOUSE = 'FULL_HOUSE'
FOUR_OF_A_KIND = 'FOUR_OF_A_KIND'
LITTLE_STRAIGHT = 'LITTLE_STRAIGHT'
BIG_STRAIGHT = 'BIG_STRAIGHT'
CHOICE = 'CHOICE'



def score(dice, category):

    match category:
        case 'YACHT':
            duplicates = [number for number in dice if dice.count(number) > 4]
            if len(duplicates) == 5:
                return 50
            else:
                return 0
        case 'ONES':
           return 1*dice.count(1)
        case 'TWOS':
           return 2*dice.count(2)
        case 'THREES':
           return 3*dice.count(3)
        case 'FOURS':
           return 4*dice.count(4)
        case 'FIVES':
           return 5*dice.count(5)
        case 'SIXES':
           return 6*dice.count(6)
        
        case 'FULL_HOUSE':
           duplicates_two = [number for number in dice if dice.count(number) == 2]
           duplicates_three = [number for number in dice if dice.count(number) == 3]

           if len(duplicates_two) == 2 and len(duplicates_three) == 3:
                return sum(dice)
           else: 
              return 0

        case 'FOUR_OF_A_KIND':
           duplicates = [number for number in dice if dice.count(number) >= 4]
           print(dice, duplicates)
           if len(duplicates) >= 4:
              return sum(duplicates[:4])
           else: 
              return 0
        case 'LITTLE_STRAIGHT':
           if set(dice) == set([1,2,3,4,5]):
                return 30
           else:
              return 0

        case 'BIG_STRAIGHT':
           if set(dice) == set([2,3,4,5,6]):
                return 30
           else:
              return 0

        case 'CHOICE':
           return sum(dice)
        case _:
           return 0

File: c/yacht/yacht.c
#include "yacht.h"

int count(int face, int faces[DICE_COUNT]) {
  int count = 0;
  for (int i = 0; i < DICE_COUNT; i++)

    if (faces[i] == face) {
      count++;
    }

  return count;
}

int sum_array(int faces[DICE_COUNT]) {
  int sum = 0;
  for (int i = 0; i < DICE_COUNT; i++)
    sum += faces[i];

  return sum;
}

int score(dice_t dice, category_t category) {

  int score = 0;
  if (ONES <= category && category <= SIXES)
    score = (category + 1) * count(category + 1, dice.faces);
  else if (category == FULL_HOUSE) {

    int focus = 0;
    for (int i = 1; i <= DICE_SIDES; i++) {

      if (count(i, dice.faces) == 3) {
        focus = i;
        break;
      }
    }

    if (focus)
      for (int i = 1; i <= DICE_SIDES; i++) {
        if (count(i, dice.faces) == 2) {
          score = sum_array(dice.faces);
          break;
        }
      }
  } else if (category == FOUR_OF_A_KIND) {
    for (int i = 1; i <= DICE_SIDES; i++)
      if (count(i, dice.faces) >= 4)
        score = i * 4;
  } else if (category == LITTLE_STRAIGHT) {
    if (count(1, dice.faces) == 1 && count(2, dice.faces) == 1 &&
        count(3, dice.faces) == 1 && count(4, dice.faces) == 1 &&
        count(5, dice.faces) == 1)
      score = 30;
  } else if (category == BIG_STRAIGHT) {
    if (count(2, dice.faces) == 1 && count(3, dice.faces) == 1 &&
        count(4, dice.faces) == 1 && count(5, dice.faces) == 1 &&
        count(6, dice.faces) == 1)
      score = 30;
  } else if (category == CHOICE)
    score = sum_array(dice.faces);
  else if (category == YACHT)
    if (count(5, dice.faces) == 5)
      score = 50;

  return score;
}
